### PR TITLE
Remainings of staking on chain logic

### PR DESCRIPTION
**Pull request by kaichaosun**

### CODE REVIEW

File: pallets/staking/src/lib.rs [hunk 1]
```diff
@@ -279,6 +299,37 @@ pub mod pallet {
...
+        ///
+        /// May only be called from `T::WithdrawOrigin`.
+        ///
+        /// - `agent`: the multisig account of relay chain.
+        /// - `amount`: the rewarded assets.
+        #[pallet::weight(10_000)]
+        #[transactional]
+        pub fn record_slash(
+            origin: OriginFor<T>,
+            agent: T::AccountId,
```
Comment (ayushmishra2005) [inline pallets/staking/src/lib.rs:313]: Can we remove this? We are not using this except sending in event.

File: pallets/staking/src/lib.rs [hunk 2]
```diff
@@ -327,6 +378,110 @@ pub mod pallet {
...
+        ///
+        /// - `agent`: the multisig account of relay chain.
+        /// - `owner`: the account which performs `unstake` operation
+        /// - `amount`: the assets can be unbond for the owner's unstaking request.
+        #[pallet::weight(10_000)]
+        #[transactional]
+        pub fn process_pending_unstake(
+            origin: OriginFor<T>,
+            agent: T::AccountId,
+            owner: T::AccountId,
```
Comment (ayushmishra2005) [inline pallets/staking/src/lib.rs:394]: Is the owner different from origin?

File: pallets/staking/src/lib.rs [hunk 3]
```diff
@@ -327,6 +378,110 @@ pub mod pallet {
...
+        ///
+        /// - `agent`: the multisig account of relay chain.
+        /// - `owner`: the account which performs `unstake` operation
+        /// - `amount`: the assets already unbond for the owner's unstaking request.
+        #[pallet::weight(10_000)]
+        #[transactional]
+        pub fn finish_processed_unstake(
+            origin: OriginFor<T>,
+            agent: T::AccountId,
+            owner: T::AccountId,
```
Comment (ayushmishra2005) [inline pallets/staking/src/lib.rs:446]: Same as above. Can we use origin as owner?

File: pallets/staking/src/lib.rs (hunk 1 above)
Comment (kaichaosun) [reply pallets/staking/src/lib.rs:313]: If no agent, there will no info to track which agent introduce slash, now we use the event to keep this info, but can be stored in storage in the future if need.

File: pallets/staking/src/lib.rs (hunk 2 above)
Comment (kaichaosun) [reply pallets/staking/src/lib.rs:394]: The origin should be the one to one map to agent, need also to be multisig.

**Review by MrPai — COMMENTED** [review]
looks good for now, I'm thinking about what if we send a msg twice?
because `stake-client` acts as an oracle that feeds data to para-chain, we should consider data duplication. I build an issue to track this. [121](https://github.com/parallel-finance/parallel/issues/121)

**Review by ayushmishra2005 — APPROVED** [review]

**Pull request merged**